ros/src/pybullet_ros/pybullet_robot.py

```diff
@@ -0,0 +1,118 @@
+import re
+
+import pybullet as pb
+import rospy
+from sensor_msgs.msg import JointState
+
+from .pybullet_robot_description import PyBulletRobotDescription
+
+
+class PyBulletRobot(PyBulletRobotDescription):
+    """
+    PyBulletRobot extending from PyBulletRobotDescription.
+    This is a collection of methods that gather 'real-time' information about the robot state and take control over the
+    robot. In other words, these are the methods called (at each time step) in the simulation loop.
+
+    Available methods (for usage, see documentation at function definition):
+        - get_joint_state_msg
+        - get_joint_state
+        - get_link_state
+        - get_base_state
+    """
+
+    def __init__(self, name, uid):
+        """
+        Constructor of the PyBulletRobot class.
+
+        :param name: the name of the robot
+        :param uid: server id of PyBullet
+
+        :type name: str
+        :type uid: int
+        """
+        assert isinstance(name, str), "[PyBulletRobot::init] Parameter 'name' has an incorrect type."
+        assert isinstance(uid, int), "[PyBulletRobot::init] Parameter 'uid' has an incorrect type."
+        self._uid = uid
+        allowed = re.compile("[a-zA-Z0-9_]*$")
+        if not allowed.match(name):
+            rospy.logerr(
+                "[PyBulletRobot::init] Invalid name '{}' for a PyBulletRobot. Allowed characters are [a-zA-Z0-9_]. " +
+                "Exiting now.".format(name))
+            return
+        self._namespace = "/" + name + "/"
+        PyBulletRobotDescription.__init__(self, self._namespace, self._uid)
+        self._initialized = self.is_initialized()
+
+    def get_joint_state_msg(self):
+        msg = JointState()
+        msg.name = [self.get_all_joint_names()[i] for i in self.get_joint_indices()]
+        msg.position, msg.velocity, _, msg.effort = self.get_joint_state()
+        return msg
+
+    def get_joint_state(self, joint_id=None):
+        """
+        Get joint state(s) (position, velocity, force, effort).
+
+        :param joint_id: Optional parameter, if different from None, then only the joint state of the desired joint is
+                         returned (if it exists), otherwise the joint states of all joints are returned.
+        :type joint_id: int
+
+        :return: Joint positions, velocities, reaction forces, and efforts of all movable joints given by bullet physics
+        :rtype: list of float
```

Comment:
isn't this rather a "4 element tuple of lists of floats", or a "4 element tuple of floats" in the case of single joint id?

Reply:
Yeah with this :rtype:, I cannot always be as precise as I'd like to because it doesn't recognize everything as return type in the generated docstring. But it would probably be more correct to have it as `tuple of float`